src/plugins/intel_gpu/src/graph/loop.cpp

```diff
@@ -316,181 +381,275 @@ void loop_inst::update_mapped_memory() {
             }
         }
     }
 }
 
 void loop_inst::preprocess_backedge_memory() {
     // checking if memory is a destination of a backedge
-    for (const auto& back_edge : _back_edges) {
+    for (size_t idx = 0; idx < _back_edges.size(); idx++) {
+        const auto& back_edge = _back_edges[idx];
         //find corresponding input of the backedge
         const auto input_map_ptrs = find_io_primitive_maps(_input_primitive_maps,
-                                                           _output_primitive_maps, back_edge.to, false);
+                                                            _output_primitive_maps, back_edge.to, false);
         const auto backedge_to_prim = body_network->get_primitive(back_edge.to);
         const auto backedge_from_prim = body_network->get_primitive(back_edge.from);
 
         memory::ptr initial_mem;
-        if (back_edge.to == _current_iteration_id) {
-            const layout current_iteration_layout = backedge_to_prim->output_memory().get_layout();
-            initial_mem = get_network().get_engine().allocate_memory(current_iteration_layout);
-            auto& stream = get_network().get_stream();
-            loop_node::write_scalar_value(initial_mem, stream, 0);
-            current_iteratoin_backedge_mapping_idx = backedge_memory_mappings.size();
-        } else {
-            if (input_map_ptrs.empty()) {
-                CLDNN_ERROR_MESSAGE(id(), "no input_mapping for backedged input");
-            }
-            initial_mem = get_external_memory(input_map_ptrs.front()->external_id);
-        }
+        OPENVINO_ASSERT(!input_map_ptrs.empty(), id(), " has no input_mapping for backedged input");
+        auto& external_id = input_map_ptrs.front()->external_id;
+        initial_mem = get_external_memory(external_id.pid, external_id.idx);
 
-        auto backedged_sliced_output_mems = get_sliced_mem(back_edge.from);
-        if (backedged_sliced_output_mems.empty()) {
+        GPU_DEBUG_LOG << idx << ") back_edge mapping - back_edge.from " << back_edge.from << std::endl;
+        GPU_DEBUG_LOG << idx << ") back_edge mapping - back_edge.to   " << back_edge.to << std::endl;
+
+        auto backedged_sliced_output = get_sliced_mem(back_edge.from);
+        if (backedged_sliced_output == nullptr) {
             // backedge output which does not need concatenation
             const auto output_mapping = find_io_primitive_maps(_input_primitive_maps,
-                                                               _output_primitive_maps, back_edge.from, false);
+                                                                _output_primitive_maps, back_edge.from, false);
             memory::ptr backedge_mem;
             if (output_mapping.empty()) {
                 // from and to primitives in backedge are connected directly
                 if (backedge_to_prim == backedge_from_prim->dependencies().front().first) {
+                    // SINGLE mode
```

Comment:
There are a lot of conditions. :) Could you please divide the conditions using mode, if it is possible.

Reply:
Fixed it. Please review it again. thanks.